File: microservice/scraper_api.py
"""
Lead Finder Microservice - REST API para scraping
Desplegado en Render.com (free tier)
Llama a este endpoint desde Vercel para ejecutar búsquedas.
"""

import os
import sys
import json
import asyncio
import re
from datetime import datetime
from supabase import create_client, Client
from fastapi import FastAPI, HTTPException
from fastapi.middleware.cors import CORSMiddleware
from pydantic import BaseModel
from typing import Optional, List
import logging

logging.basicConfig(level=logging.INFO)
logger = logging.getLogger(__name__)

# ─── Supabase Config ───
SUPABASE_URL = os.environ.get("SUPABASE_URL", "")
SUPABASE_KEY = os.environ.get("SUPABASE_SERVICE_KEY", "")

if not SUPABASE_URL or not SUPABASE_KEY:
    # Try to load from env file for local testing
    try:
        with open(".env.local") as f:
            for line in f:
                if "=" in line:
                    k, v = line.strip().split("=", 1)
                    os.environ.setdefault(k, v)
        SUPABASE_URL = os.environ.get("SUPABASE_URL", "")
        SUPABASE_KEY = os.environ.get("SUPABASE_SERVICE_KEY", "")
    except:
        pass

supabase: Optional[Client] = None
if SUPABASE_URL and SUPABASE_KEY:
    supabase = create_client(SUPABASE_URL, SUPABASE_KEY)

# ─── FastAPI App ───
app = FastAPI(title="Lead Finder Microservice")

app.add_middleware(
    CORSMiddleware,
    allow_origins=["*"],
    allow_credentials=True,
    allow_methods=["*"],
    allow_headers=["*"],
)

# ─── Models ───
class SearchRequest(BaseModel):
    category: str
    state: str
    city: str
    parish: Optional[str] = None
    sector: Optional[str] = None
    deep: bool = True
    googleSearch: bool = False
    paginasAmarillas: bool = False
    social: bool = False
    tiktok: bool = False
    instagram: bool = False
    maxDeep: Optional[int] = None

class SearchResponse(BaseModel):
    jobId: str
    status: str
    leadsFound: Optional[int] = None
    message: str

# ─── Helpers ───
def get_location(state: str, city: str) -> str:
    """Construct location string for API queries"""
    return f"{city}, {state}, Venezuela"

def slugify(text: str) -> str:
    return text.lower().replace(" ", "_").replace("ñ", "n").replace("á", "a").replace("é", "e").replace("í", "i").replace("ó", "o").replace("ú", "u")

def clean_phone(phone_str: str) -> str:
    """Clean and validate Venezuelan phone numbers"""
    if not phone_str:
        return ""
    phone = re.sub(r'[^\d+]', '', str(phone_str))
    if phone.startswith("58") and len(phone) > 10:
        return f"+{phone}"
    if len(phone) == 10 and phone.startswith("412"):
        return f"+58{phone}"
    if len(phone) == 11 and phone.startswith("0412"):
        return f"+58{phone[1:]}"
    return phone if phone else ""

def clean_url(url: str) -> str:
    if not url:
        return ""
    url = str(url).strip()
    if url and not url.startswith(("http://", "https://")):
        url = "https://" + url
    return url

async def scrape_google_maps(category: str, state: str, city: str, deep: bool = True, max_deep: int = 5) -> List[dict]:
    """Scrape Google Maps using the maps.py spider logic"""
    try:
        from scrapling import StealthySession
        
        location = get_location(state, city)
        search_url = f"https://www.google.com/maps/search/{category}+en+{location.replace(' ', '+')}/"
        
        logger.info(f"Starting scrape: {search_url}")
        
        session = StealthySession()
        session.set_option('timeout', 30000)
        session.set_option('retry_limit', 3)
        
        page = session.get(search_url)
        
        # Handle cookie consent
        try:
            page.evaluate('document.querySelector("button[aria-label*=\'Aceptar\']")?.click()')
            page.wait(500)
        except:
            pass
        
        # Scroll to load more results
        for _ in range(20):
            try:
                page.evaluate('document.querySelectorAll(".fontHeadlineSmall")[0]?.scrollIntoView()')
                page.wait(600)
            except:
                break
        
        # Extract place cards
        places = page.evaluate('''
            () => {
                const cards = document.querySelectorAll('[data-categorical-id]');
                return Array.from(cards).map(card => {
                    const nameEl = card.querySelector('.fontHeadlineSmall') || card.querySelector('.qBF1Pd');
                    const ratingEl = card.querySelector('.ZkPwu') || card.querySelector('[aria-label*="estrel"]');
                    const reviewsEl = card.querySelector('.TyH0Cd') || card.querySelector('.ODfHxe');
                    const addressEl = card.querySelector('.W4VJmc') || card.querySelector('.fontBodyMedium');
                    const phoneEl = card.querySelector('.USJqEc') || card.querySelector('.CFH2De');
                    const categoryEl = card.querySelector('.dSVPB') || card.querySelector('.wNXCwl');
                    
                    const name = nameEl?.textContent?.trim() || '';
                    const ratingStr = ratingEl?.getAttribute('aria-label') || ratingEl?.textContent || '';
                    const rating = ratingStr.includes('estrel') ? parseFloat(ratingStr.replace(/[^0-9.,]/g, '').replace(',', '.')) : null;
                    const reviewsStr = reviewsEl?.textContent?.replace(/[^0-9]/g, '') || '0';
                    const reviews = parseInt(reviewsStr) || 0;
                    const address = addressEl?.textContent?.trim() || '';
                    const phone = phoneEl?.textContent?.trim() || '';
                    const category = categoryEl?.textContent?.trim() || '';
                    
                    const link = card.querySelector('a')?.href || '';
                    const isAd = !!card.querySelector('.SiBjxf') || !!card.querySelector('[data-ad]');
                    
                    return { name, rating, reviews, address, phone, category, link, isAd };
                }).filter(p => p.name && !p.isAd);
            }
        ''')
        
        logger.info(f"Found {len(places)} places, deep={deep}")
        
        leads = []
        seen = set()
        
        for place in places[:30]:
            if place['name'] in seen:
                continue
            seen.add(place['name'])
            
            lead = {
                'name': place['name'],
                'category': category,
                'location': f"{city}, {state}",
                'state': state,
                'city': city,
                'address': place.get('address', '') or '',
                'phone': clean_phone(place.get('phone', '')),
                'rating': place.get('rating'),
                'reviews_count': place.get('reviews', 0),
                'source': 'google_maps',
                'source_url': place.get('link', ''),
                'status': 'frio',
                'timestamp': datetime.now().isostring(),
                'notes': '',
            }
            
            if deep and place.get('link'):
                try:
                    detail_page = session.get(place['link'])
                    detail_page.wait(1000)
                    
                    website = detail_page.evaluate('''
                        () => {
                            const websiteLink = document.querySelector('a[data-item-id="authority"]') 
                                || document.querySelector('a[aria-label*="sitio web"]')
                                || document.querySelector('a[href*="://"]');
                            return websiteLink?.href || '';
                        }
                    ''')
                    
                    email = detail_page.evaluate('''
                        () => {
                            const emailLink = document.querySelector('a[href^="mailto:"]');
                            return emailLink?.href?.replace('mailto:', '') || '';
                        }
                    ''')
                    
                    lead['website'] = clean_url(website)
                    lead['email'] = email or ''
                    
                    # Social links
                    page_text = detail_page.text or ''
                    fb = re.search(r'facebook\.com/[\w-]+', page_text)
                    ig = re.search(r'instagram\.com/[\w-]+', page_text)
                    tw = re.search(r'twitter\.com/[\w-]+|x\.com/[\w-]+', page_text)
                    
                    lead['facebook'] = f"https://{fb.group(0)}" if fb else ''
                    lead['instagram'] = f"https://{ig.group(0)}" if ig else ''
                    lead['twitter'] = f"https://{tw.group(0)}" if tw else ''
                    
                    logger.info(f"  Deep data for: {place['name']}")
                    
                except Exception as e:
                    logger.warning(f"  Deep scrape failed for {place['name']}: {e}")
            
            leads.append(lead)
        
        return leads
        
    except Exception as e:
        logger.error(f"Scraping error: {e}")
        return []

async def save_leads_to_supabase(leads: List[dict]) -> int:
    """Save leads to Supabase, skip duplicates"""
    if not supabase or not leads:
        return 0
    
    saved = 0
    for lead in leads:
        try:
            # Check for existing lead with same name + address
            existing = supabase.from_('leads').select('id').eq('name', lead['name']).eq('address', lead.get('address', '')).execute()
            
            if existing.data:
                continue  # Skip duplicate
            
            data = {k: v for k, v in lead.items() if v is not None and v != ''}
            data['rating'] = data.get('rating') or None
            data['reviews_count'] = data.get('reviews_count') or 0
            
            supabase.from_('leads').insert(data).execute()
            saved += 1
            logger.info(f"Saved: {lead['name']}")
        except Exception as e:
            logger.warning(f"Failed to save {lead['name']}: {e}")
    
    return saved

# ─── Routes ───
@app.get("/")
async def root():
    return {"status": "ok", "service": "Lead Finder Microservice", "time": datetime.now().isostring()}

@app.get("/health")
async def health():
    has_supabase = supabase is not None
    return {
        "status": "healthy",
        "supabase_connected": has_supabase,
        "time": datetime.now().isostring()
    }

@app.post("/search", response_model=SearchResponse)
async def search(request: SearchRequest):
    """Execute a lead search and save results to Supabase"""
    
    if not supabase:
        raise HTTPException(status_code=500, detail="Supabase no configurado")
    
    job_id = f"search_{datetime.now().strftime('%Y%m%d_%H%M%S')}"
    
    logger.info(f"Starting job {job_id}: {request.category} in {request.city}, {request.state}")
    
    try:
        # Run the scrape
        leads = await scrape_google_maps(
            category=request.category,
            state=request.state,
            city=request.city,
            deep=request.deep,
            max_deep=request.maxDeep or 5,
        )
        
        # Save to Supabase
        saved_count = await save_leads_to_supabase(leads)
        
        logger.info(f"Job {job_id} completed: {saved_count} new leads saved")
        
        return SearchResponse(
            jobId=job_id,
            status="done",
            leadsFound=saved_count,
            message=f"{saved_count} leads guardados en {request.city}, {request.state}"
        )
        
    except Exception as e:
        logger.error(f"Job {job_id} failed: {e}")
        raise HTTPException(status_code=500, detail=str(e))

@app.get("/stats")
async def stats():
    """Return current stats from Supabase"""
    if not supabase:
        raise HTTPException(status_code=500, detail="Supabase no configurado")
    
    try:
        result = supabase.from_('leads').select("status, category, state", count="exact").execute()
        total = len(result.data)
        by_status = {}
        by_category = {}
        by_state = {}
        
        for lead in result.data:
            s = lead.get('status', 'frio')
            by_status[s] = by_status.get(s, 0) + 1
            cat = lead.get('category', 'unknown')
            by_category[cat] = by_category.get(cat, 0) + 1
            st = lead.get('state', 'unknown')
            by_state[st] = by_state.get(st, 0) + 1
        
        return {"total": total, "byStatus": by_status, "byCategory": by_category, "byState": by_state}
    except Exception as e:
        logger.error(f"Stats error: {e}")
        raise HTTPException(status_code=500, detail=str(e))

if __name__ == "__main__":
    import uvicorn
    port = int(os.environ.get("PORT", 8000))
    uvicorn.run(app, host="0.0.0.0", port=port)

File: microservice_js/scraper_api.js
/**
 * Lead Finder Microservice - Node.js + Playwright
 * Desplegado en Render.com (free tier)
 * Unext.js dashboard en Vercel llama a este endpoint para ejecutar búsquedas.
 */

const express = require('express');
const cors = require('cors');
const { chromium } = require('playwright');
const { createClient } = require('@supabase/supabase-js');
require('dotenv').config();

const app = express();
app.use(cors());
app.use(express.json());

// ─── Supabase Config ───
const SUPABASE_URL = process.env.SUPABASE_URL || '';
const SUPABASE_KEY = process.env.SUPABASE_SERVICE_KEY || '';

let supabase = null;
if (SUPABASE_URL && SUPABASE_KEY) {
  supabase = createClient(SUPABASE_URL, SUPABASE_KEY);
  console.log('[OK] Supabase conectado');
} else {
  console.log('[WARN] Supabase no configurado - guardando solo en JSON local');
}

// ─── Helpers ───
function getLocation(state, city) {
  return `${city}, ${state}, Venezuela`;
}

function slugify(text) {
  return text.toLowerCase()
    .replace(/ /g, '_')
    .replace(/ñ/g, 'n')
    .replace(/á/g, 'a')
    .replace(/é/g, 'e')
    .replace(/í/g, 'i')
    .replace(/ó/g, 'o')
    .replace(/ú/g, 'u');
}

function cleanPhone(phoneStr) {
  if (!phoneStr) return '';
  let phone = phoneStr.replace(/[^\d+]/g, '');
  if (phone.startsWith('58') && phone.length > 10) return `+${phone}`;
  if (phone.length === 10 && phone.startsWith('412')) return `+58${phone}`;
  if (phone.length === 11 && phone.startsWith('0412')) return `+58${phone.substring(1)}`;
  return phone || '';
}

function cleanUrl(url) {
  if (!url) return '';
  url = url.trim();
  if (url && !url.startsWith('http://') && !url.startsWith('https://')) {
    url = 'https://' + url;
  }
  return url;
}

// ─── Scraper ───
async function scrapeGoogleMaps(category, state, city, deep = true, maxDeep = 5) {
  const location = getLocation(state, city);
  const searchUrl = `https://www.google.com/maps/search/${encodeURIComponent(category + ' en ' + location).replace(/%20/g, '+')}`;

  console.log(`[SCRAPER] Navegando a: ${searchUrl}`);

  const browser = await chromium.launch({
    headless: true,
    args: ['--no-sandbox', '--disable-setuid-sandbox']
  });

  const context = await browser.newContext({
    locale: 'es-ES',
    userAgent: 'Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/120.0.0.0 Safari/537.36'
  });

  const page = await context.newPage();

  // Bloquear recursos innecesarios para acelerar
  await context.route('**/*.{png,jpg,jpeg,gif,css,woff,woff2,font}', route => route.abort());

  try {
    await page.goto(searchUrl, { waitUntil: 'domcontentloaded', timeout: 30000 });
    console.log('[SCRAPER] Página cargada');

    // Aceptar cookies si aparece el popup
    try {
      const acceptBtn = page.locator('button[aria-label*="Aceptar"], button[aria-label*="Accept"]').first();
      if (await acceptBtn.isVisible({ timeout: 3000 })) {
        await acceptBtn.click();
        console.log('[SCRAPER] Popup de cookies aceptado');
        await page.waitForTimeout(2000);
      }
    } catch (e) {
      // No hay popup, continuar
    }

    // Scroll para cargar más resultados
    console.log('[SCRAPER] Haciendo scroll para cargar resultados...');
    for (let i = 0; i < 15; i++) {
      try {
        await page.evaluate(() => {
          const feed = document.querySelector('[role="feed"]');
          if (feed) {
            feed.scrollBy(0, feed.clientHeight || 500);
          } else {
            window.scrollBy(0, window.innerHeight || 500);
          }
        });
        await page.waitForTimeout(800);
      } catch (e) {
        break;
      }
    }

    // Extraer datos de las tarjetas
    const places = await page.evaluate(() => {
      const feed = document.querySelector('[role="feed"]');
      if (!feed) return [];

      const articles = feed.querySelectorAll('div[role="article"]');
      const results = [];
      const seen = new Set();

      articles.forEach(article => {
        if (seen.has(article)) return;
        seen.add(article);

        let name = article.getAttribute('aria-label') || '';
        if (!name || name.length < 2) {
          const ne = article.querySelector('.fontHeadlineSmall, [role="heading"], h1, h2, h3');
          if (ne) name = (ne.textContent || '').trim();
        }
        if (!name || name.length < 2) return;

        let rating = null, reviews = null;
        const reImg = article.querySelector('[role="img"][aria-label]');
        if (reImg) {
          const lbl = reImg.getAttribute('aria-label') || '';
          const rm = lbl.match(/([\d.,]+)/);
          if (rm) rating = parseFloat(rm[1].replace(',', '.'));
          const rvm = lbl.match(/\((\d+)\s*/);
          if (rvm) reviews = parseInt(rvm[1]);
        }

        let phone = '', website = '';
        article.querySelectorAll('a[href]').forEach(a => {
          const h = a.href || '';
          if (h.startsWith('tel:')) phone = h.replace('tel:', '').split('?')[0];
          else if (h.startsWith('http') && !h.includes('google.com/maps') && !h.includes('google.com/search') && !website) website = h;
        });

        // Extraer dirección de las secciones
        const sections = article.querySelectorAll('.W4Efsd, .fontBodyMedium');
        let address = '';
        const allLines = [];
        sections.forEach(sec => {
          const txt = sec.textContentContent || sec.textContent || '';
          const lines = txt.split('\n').map(s => s.trim()).filter(Boolean);
          lines.forEach(l => allLines.push(l));
        });

        // Buscar línea que parezca dirección
        const addrPatterns = ['Calle', 'Av.', 'Avenida', 'Carretera', 'Urb.', 'Urbanizacion', 'Carrera', 'Edif.', 'Edificio', 'Centro Comercial', 'CC ', 'Local ', 'Sector ', 'Barrio ', 'Torre ', 'Km ', 'Esq.', 'Transversal', 'Via ', 'Boulevard'];
        for (const line of allLines) {
          if (line !== name && line.length > 5 && line.length < 200) {
            if (addrPatterns.some(p => line.includes(p))) {
              address = line;
              break;
            }
          }
        }

        if (!address) {
          for (const line of allLines) {
            if (line !== name && line.length > 5 && line.length < 200 && /^[A-ZÁÉÍÓÚÑ][a-záéíóúñ]+\s+\d/.test(line)) {
              address = line;
              break;
            }
          }
        }

        const linkEl = article.querySelector('a[href*="/maps/place/"]');
        const href = linkEl ? linkEl.href : '';

        results.push({ name, rating, reviews, address, phone, website, href });
      });

      return results;
    });

    console.log(`[SCRAPER] Encontrados ${places.length} lugares`);

    const leads = [];
    const seenNames = new Set();

    for (const place of places.slice(0, 30)) {
      if (seenNames.has(place.name)) continue;
      seenNames.add(place.name);

      const lead = {
        name: place.name,
        category: category,
        location: `${city}, ${state}`,
        state: state,
        city: city,
        address: place.address || '',
        phone: cleanPhone(place.phone),
        website: cleanUrl(place.website),
        rating: place.rating,
        reviews_count: place.reviews || 0,
        source: 'google_maps',
        source_url: place.href,
        status: 'frio',
        timestamp: new Date().toISOString(),
        notes: ''
      };

      // Deep scraping: obtener más datos si está habilitado
      if (deep && place.href && lead.length < maxDeep) {
        try {
          await page.goto(place.href, { waitUntil: 'domcontentloaded', timeout: 15000 });
          await page.waitForTimeout(1500);

          // Aceptar cookies en la página del lugar
          try {
            const acceptBtn = page.locator('button[aria-label*="Aceptar"], button[aria-label*="Accept"]').first();
            if (await acceptBtn.isVisible({ timeout: 2000 })) {
              await acceptBtn.click();
              await page.waitForTimeout(1000);
            }
          } catch (e) {}

          // Extraer website, email y redes sociales
          const deepData = await page.evaluate(() => {
            let website = '', email = '', facebook = '', instagram = '', twitter = '';
            const pageText = document.body ? document.body.innerText : '';

            const websiteLink = document.querySelector('a[data-item-id="authority"]') || document.querySelector('a[aria-label*="sitio web"]') || document.querySelector('a[href*="://"]:not([href*="google.com"]):not([href*="maps"])');
            if (websiteLink) website = websiteLink.href || '';

            const emailLink = document.querySelector('a[href^="mailto:"]');
            if (emailLink) email = emailLink.href.replace('mailto:', '') || '';

            const fbMatch = pageText.match(/facebook\.com\/[\w-]+/i);
            const igMatch = pageText.match(/instagram\.com\/[\w-]+/i);
            const twMatch = pageText.match(/twitter\.com\/[\w-]+|x\.com\/[\w-]+/i);

            if (fbMatch) facebook = 'https://' + fbMatch[0];
            if (igMatch) instagram = 'https://' + igMatch[0];
            if (twMatch) twitter = 'https://' + twMatch[0];

            return { website, email, facebook, instagram, twitter };
          });

          lead.website = cleanUrl(deepData.website || lead.website);
          lead.email = deepData.email || '';
          lead.facebook = deepData.facebook || '';
          lead.instagram = deepData.instagram || '';
          lead.twitter = deepData.twitter || '';

          console.log(`[SCRAPER]   Deep data: ${place.name}`);
        } catch (e) {
          console.log(`[SCRAPER]   Error en deep scrape: ${e.message}`);
        }
      }

      leads.push(lead);
    }

    await browser.close();
    return leads;

  } catch (error) {
    console.error('[SCRAPER ERROR]', error.message);
    await browser.close();
    return [];
  }
}

// ─── Save to Supabase ───
async function saveLeadsToSupabase(leads) {
  if (!supabase || !leads || leads.length === 0) {
    console.log('[DB] Sin Supabase o sin leads, guardando en JSON local...');
    const fs = require('fs');
    const path = require('path');
    const jsonPath = path.join(__dirname, '..', 'leads_data', 'leads.json');

    let existing = [];
    try {
      if (fs.existsSync(jsonPath)) {
        existing = JSON.parse(fs.readFileSync(jsonPath, 'utf8'));
      }
    } catch (e) {}

    const newLeads = leads.filter(l => !existing.some(e => e.name === l.name && e.address === l.address));
    existing.push(...newLeads);

    fs.mkdirSync(path.dirname(jsonPath), { recursive: true });
    fs.writeFileSync(jsonPath, JSON.stringify(existing, null, 2));

    return newLeads.length;
  }

  let saved = 0;
  for (const lead of leads) {
    try {
      const { data: existing } = await supabase
        .from('leads')
        .select('id')
        .eq('name', lead.name)
        .eq('address', lead.address || '')
        .limit(1);

      if (existing && existing.length > 0) continue;

      const cleanLead = {};
      Object.keys(lead).forEach(k => {
        if (lead[k] !== null && lead[k] !== undefined && lead[k] !== '') {
          cleanLead[k] = lead[k];
        }
      });
      cleanLead.rating = cleanLead.rating || null;
      cleanLead.reviews_count = cleanLead.reviews_count || 0;

      await supabase.from('leads').insert(cleanLead);
      saved++;
      console.log(`[DB] Guardado: ${lead.name}`);
    } catch (e) {
      console.log(`[DB] Error guardando ${lead.name}: ${e.message}`);
    }
  }
  return saved;
}

// ─── Routes ───
app.get('/', (req, res) => {
  res.json({
    status: 'ok',
    service: 'Lead Finder Microservice (Node.js + Playwright)',
    time: new Date().toISOString()
  });
});

app.get('/health', (req, res) => {
  res.json({
    status: 'healthy',
    supabase_connected: !!supabase,
    time: new Date().toISOString()
  });
});

app.post('/search', async (req, res) => {
  const { category, state, city, parish, sector, deep, googleSearch, paginasAmarillas, social, tiktok, instagram, maxDeep } = req.body;

  if (!category || !state || !city) {
    return res.status(400).json({ error: 'category, state, and city are required' });
  }

  console.log(`\n[SEARCH] ${category} en ${city}, ${state} (deep=${deep})`);

  try {
    const leads = await scrapeGoogleMaps(category, state, city, deep !== false, maxDeep || 5);
    const savedCount = await saveLeadsToSupabase(leads);

    const jobId = `search_${Date.now()}`;
    console.log(`[SEARCH] Completado: ${savedCount} nuevos leads`);

    res.json({
      jobId,
      status: 'done',
      leadsFound: savedCount,
      message: `${savedCount} leads guardados en ${city}, ${state}`
    });
  } catch (error) {
    console.error('[SEARCH ERROR]', error);
    res.status(500).json({ error: error.message });
  }
});

app.get('/stats', async (req, res) => {
  if (!supabase) {
    return res.status(500).json({ error: 'Supabase no configurado' });
  }

  try {
    const { data, count } = await supabase
      .from('leads')
      .select('status, category, state', { count: 'exact' });

    const total = count || (data ? data.length : 0);
    const byStatus = {}, byCategory = {}, byState = {};

    if (data) {
      data.forEach(lead => {
        const s = lead.status || 'frio';
        byStatus[s] = (byStatus[s] || 0) + 1;
        const cat = lead.category || 'unknown';
        byCategory[cat] = (byCategory[cat] || 0) + 1;
        const st = lead.state || 'unknown';
        byState[st] = (byState[st] || 0) + 1;
      });
    }

    res.json({ total, byStatus, byCategory, byState });
  } catch (error) {
    res.status(500).json({ error: error.message });
  }
});

// ─── Start ───
const PORT = process.env.PORT || 3001;
app.listen(PORT, () => {
  console.log(`\n🚀 Lead Finder Microservice iniciado en puerto ${PORT}`);
  console.log(`   Entorno: ${process.env.NODE_ENV || 'development'}`);
  console.log(`   Supabase: ${supabase ? 'Conectado' : 'No configurado'}\n`);
});

module.exports = app;